### PR TITLE
Control

### PR DESCRIPTION
**Pull request by Stapelzeiger**
mostly works

### CODE REVIEW

**Comment by Stapelzeiger** [comment]
review anyone?

File: src/control.c [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+struct feedback_s control_feedback;
+motor_protection_t control_motor_protection;
+
+
+static struct pid_cascade_s ctrl;
+
+static bool motor_enable = true;
+
+#define SETPT_MODE_POS      0
```
Comment (antoinealb) [inline src/control.c:29]: Maybe an enum would be better here.

File: src/control.c [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                            parameter_scalar_get(&p->kd));
+        pid_reset_integral(ctrl);
+    }
+    if (parameter_changed(&p->i_limit)) {
+        pid_set_integral_limit(ctrl, parameter_scalar_get(&p->i_limit));
+    }
+}
+
+// control loop parameters
+static parameter_namespace_t param_ns_control;
```
Comment (antoinealb) [inline src/control.c:209]: I feel like all global vars declaration should be at the top of the file.

File: src/control.c [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+            feedback_compute(&control_feedback);
+
+            ctrl.position = control_feedback.output.position;
+            ctrl.velocity = control_feedback.output.velocity;
+            ctrl.current = analog_get_motor_current();
+
+            // ctrl.current_limit = motor_protection_update(&control_motor_protection, ctrl.current, delta_t);
+
+            // setpoints
+            if (setpt_mode == SETPT_MODE_TRAJ) {
```
Comment (antoinealb) [inline src/control.c:352]: Maybe a switch case would be better ?

File: src/encoder.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+uint32_t encoder_get_secondary(void)
+{
+    return STM32_TIM3->CNT;
+}
+
+float encoder_get_speed(void)
+{
+    return encoder_speed;
```
Comment (antoinealb) [inline src/encoder.c:22]: So we assume that float copying is atomic ? How reasonable is that ?

File: src/encoder.c [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    encoder_new = encoder_get_primary();
+
+    while (42) {
+        encoder_old = encoder_new;
+        encoder_new = encoder_get_primary();
+        delta_time = chVTGetSystemTimeX() - time;
+        delta_time = 20;
+        time += delta_time;
+
+        // encoder_speed = filter_iir_apply(&speed_filter,
```
Comment (antoinealb) [inline src/encoder.c:54]: Please remove dead code.

File: src/main.c [hunk 6]
```diff
@@ -1,14 +1,208 @@
...
+
+
+void panic_hook(const char* reason)
+{
+    palClearPad(GPIOA, GPIOA_LED);      // turn on LED (active low)
+    static BlockingUARTDriver blocking_uart_stream;
+    blocking_uart_init(&blocking_uart_stream, USART3, 115200);
+    BaseSequentialStream* uart = (BaseSequentialStream*)&blocking_uart_stream;
+    int i;
+    while(42){
```
Comment (antoinealb) [inline src/main.c:94]: coding style: space after "while" and "for" (minor issue)

File: src/main.c [hunk 7]
```diff
@@ -1,14 +1,208 @@
...
+        }
+        chprintf(uart, "Panic: %s\n", reason);
+    }
+}
+
+
+static int error_level = 0;
+
+static THD_WORKING_AREA(led_thread_wa, 128);
+static THD_FUNCTION(led_thread, arg)
```
Comment (antoinealb) [inline src/main.c:106]: What does this thread do ? It is not directly obvious from the code, so a small comment would be good.

File: src/motor_protection.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct {
+    float t;
+    float t_max;
+    float t_ambient;
+    float r_th;
```
Comment (antoinealb) [inline src/motor_protection.h:12]: What are those coefficient ? I am guessing thermal capacity and thermal resistance, but maybe a comment with the unit too would be good.

**Comment by antoinealb** [comment]
The implementation looks good from a quick review. I noted a few style issues though (see line comments)

**Comment by antoinealb** [comment]
Ping for style fix.

**Pull request closed without merging**